### PR TITLE
[Stage 1] Binder HTTP

### PR DESCRIPTION
**Pull request by michlee1337**
- template class HttpBinder<bool IsReq> inherits from template specialization of Binder
- implements Binder virtual funcs
- HttpReqBinder and HttpResBinder inherits from HttpBinder specialized to IsReq=true and isReq=false (matching boost beast representation)

- [x] Tests pass
- [x] Appropriate changes to README are included in PR - User instructions in binder-pubsub branch

### CODE REVIEW

File: v1/protocol_binding/http_binder.cc [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+typedef boost::beast::http::request<boost::beast::http::string_body> HttpRequest;
+typedef absl::flat_hash_map<std::string, CloudEvent_CloudEventAttribute> CeAttrMap;
+
+constexpr absl::string_view kHttpContentKey = "Content-Type";
+
+template <>
+absl::StatusOr<std::string> Binder<HttpRequest>::GetContentType(HttpRequest& http_req) {
+    auto iter = http_req.base().find(kHttpContentKey.data());
+    if (iter == http_req.base().end()) {
+        return std::string("");
```
Comment (ashley-sheng) [inline v1/protocol_binding/http_binder.cc:23]: would `return "";` work?

File: v1/protocol_binding/http_binder.cc [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+absl::StatusOr<CloudEvent> Binder<HttpRequest>::UnbindBinary(HttpRequest& http_req) {
+    CloudEvent cloud_event;
+    for (auto it = http_req.base().begin(); it!=http_req.base().end(); ++it) {
+        std::string key;
+        std::string header_key = (*it).name_string().to_string();
+        if (header_key == kHttpContentKey.data()) {
+            key = kContenttypeKey.data();
+        } 
+        else if (header_key.rfind(kMetadataPrefix.data(), 0) == 0){
+            size_t len_prefix = strlen(kMetadataPrefix.data());
```
Comment (ashley-sheng) [inline v1/protocol_binding/http_binder.cc:43]:
```
Strlen Returns the length of the C string str.

The length of a C string is determined by the terminating null-character: A C string is as long as the number of characters between the beginning of the string and the terminating null character (without including the terminating null character itself).
```

absl::string_view doesn't guarantee that data will null terminated (implementation defined) therefore you can't expect strlen to work correctly. Please use absl::string_view::size.

File: v1/util/cloud_events_util.cc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        return absl::InvalidArgumentError("GetMetadata can only be called with valid Cloud Event.");
+    }
+
+    // create absl::flat_hash_map from protobuf map of optional/ extensionattrs
+    absl::flat_hash_map<std::string, CloudEvent_CloudEventAttribute> attrs(
+        (cloud_event.attributes()).begin(),
+        cloud_event.attributes().end());
+
+    // insert required attrs
+    CloudEvent_CloudEventAttribute attr_val;
```
Comment (ashley-sheng) [inline v1/util/cloud_events_util.cc:31]: is it outdated? I have seen you using create CeString function.

File: v1/protocol_binding/http_binder.cc (hunk 1 above)
Comment (michlee1337) [reply v1/protocol_binding/http_binder.cc]: I get a conversion error, I think with absl::StatusOr I always need explicit type declaration if it's implicit/ polymorphic

```
error: could not convert '""' const char [1]' to 'absl::StatusOr<std::__cxx11::basic_string<char> 
```

Comment (ashley-sheng) [reply v1/protocol_binding/http_binder.cc]: I see. Thanks.

File: v1/protocol_binding/http_binder.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+//
+// User interacts with HttpReqBinder and HttpResBinder
+// which are both child classes of HttpBinder.
+
+template <bool IsReq>
+class HttpBinder: public Binder<
+  boost::beast::http::message<IsReq, boost::beast::http::string_body>> {
+ private:
+  // _____ Operations used in Bind Binary _____
+
```
Comment (ashley-sheng) [inline v1/protocol_binding/http_binder.h:27]: remove this line, here and below.

**Review by ashley-sheng — APPROVED** [review]

**Review by hrasadi — APPROVED** [review]

**Pull request merged**